### PR TITLE
kirkwood: add support for Iomega Storcenter ix4-200d

### PR DESCRIPTION
**Pull request by sanderpqr**
Iomega Storcenter ix4-200d is a four-bay SATA NAS powered by a Marvell
 Kirkwood SoC clocked at 1.2GHz. It has 512MB of RAM and 32MB of
 flash memory, 3x USB 2.0 and 2x 1Gbit/s NIC

Specification:
- SoC: Marvell Kirkwood 88F6281
- CPU/Speed: 1200Mhz
- Flash size: 32 MiB
- RAM: 512MB
- LAN: 2x 1Gbit/s
- 3x USB 2.0

Notes:
- The blue drive LED is triggered by HDD activity, it can not be controlled via GPIO.
- The LCD screen requires proprietary code and does not function at this time.
- Due to a kernel-related issue with the Marvell 88SE6121 SATA controller, currently only trays numbered 3 and 4 work, 1 and 2 do not. [1]

Serial pinout:

    CN4
    --------------
    | 10 8 6 4 2 |
    |  9 7 5 3 1 |
    -------------- PIN 1 Mark (fat line)

     1 = RXD
     4 = TXD
     6 = GND
     9 = 3.3V (not necessary to connect)

Installation instructions:
1. download initramfs-uImage and copy into tftp server
2. connect the tftp server to network port 1
3. access uboot environment with serial cable and run 
    ```
    setenv mainlineLinux yes
    setenv arcNumber 1682
    setenv console 'console=ttyS0,115200n8'
    setenv mtdparts 'mtdparts=orion_nand:0x100000@0x000000(u-boot)ro,0x20000@0xA0000(u-boot environment)ro,0x300000@0x100000(kernel),0x1C00000@0x400000(ubi)'
    setenv bootargs_root 'root='
    setenv bootcmd 'setenv bootargs ${console} ${mtdparts} ${bootargs_root}; nand read.e 0x800000 0x100000 0x300000; bootm 0x00800000'
    saveenv
    setenv serverip 192.168.1.1
    setenv ipaddr 192.168.1.2
    tftpboot 0x00800000 [initramfs-uImage filename]
    bootm 0x00800000
    ```
4. connect to LAN on network port 2, log into openwrt and sysupgrade to install into flash

[1] https://bugzilla.kernel.org/show_bug.cgi?id=216094

### CODE REVIEW

File: target/linux/kirkwood/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -22,6 +22,9 @@ kirkwood_setup_interfaces()
 	globalscale,sheevaplug|\
 	iom,iconnect-1.1|\
 	iom,ix2-200|\
+	iom,ix4-200d)
+		ucidef_set_interfaces_lan_wan "eth0" "eth1"
+		;;
 	iptime,nas1|\
```
Comment (obsy) [inline target/linux/kirkwood/base-files/etc/board.d/02_network:28]: This broke other devices like sheevaplug, iconnect or ix2-200.

**Review by obsy — CHANGES_REQUESTED** [review]

File: target/linux/kirkwood/base-files/etc/board.d/02_network (hunk 1 above)
Comment (obsy) [reply target/linux/kirkwood/base-files/etc/board.d/02_network:28]: Oryginal:
```
	cloudengines,pogoe02|\
	cloudengines,pogoplugv4|\
	ctera,c200-v1|\
	globalscale,sheevaplug|\
	iom,iconnect-1.1|\
	iom,ix2-200|\
	iptime,nas1|\
	netgear,readynas-duo-v2|\
	raidsonic,ib-nas62x0|\
	seagate,blackarmor-nas220|\
	seagate,dockstar|\
	seagate,goflexhome|\
	seagate,goflexnet|\
	zyxel,nsa310b|\
	zyxel,nsa310s|\
	zyxel,nsa325)
		ucidef_set_interface_lan "eth0" "dhcp"
		;;
```

So your code should be
```
	cloudengines,pogoe02|\
	cloudengines,pogoplugv4|\
	ctera,c200-v1|\
	globalscale,sheevaplug|\
	iom,iconnect-1.1|\
	iom,ix2-200|\
	iptime,nas1|\
	netgear,readynas-duo-v2|\
	raidsonic,ib-nas62x0|\
	seagate,blackarmor-nas220|\
	seagate,dockstar|\
	seagate,goflexhome|\
	seagate,goflexnet|\
	zyxel,nsa310b|\
	zyxel,nsa310s|\
	zyxel,nsa325)
		ucidef_set_interface_lan "eth0" "dhcp"
		;;
       iom,ix4-200d)
		ucidef_set_interfaces_lan_wan "eth0" "eth1"
		;;
```
Not inside existing statement.

Comment (obsy) [reply target/linux/kirkwood/base-files/etc/board.d/02_network:28]: No. this is "normal", not inside. 
If you have
```
case $var in
a|\
b)
  echo "1"
  ;;
c)
 echo "2'
 ;;
esac
```

code shows "1" for a or b. but if you change (add) "inside":
```
case $var in
a|\
d)
  echo "0"
  ;;
b)
  echo "1"
  ;;
c)
 echo "2'
 ;;
esac
```

Then for "a" shows 0, not 1, as previous. So you code should not break existing behaviour.

File: target/linux/kirkwood/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -37,6 +34,9 @@ kirkwood_setup_interfaces()
 	zyxel,nsa325)
 		ucidef_set_interface_lan "eth0" "dhcp"
 		;;
+	iom,ix4-200d)
```
Comment (obsy) [inline target/linux/kirkwood/base-files/etc/board.d/02_network:37]: Please keep the alphabetical order.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+			pcie@1,0 {
+			status = "okay";
+			};
+		};
+};
+
+ocp@f1000000 {
+	pinctrl: pin-controller@10000 {
+		spi_pin1: spi-pin1 {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:33]: Shrink pin properties into one is easier to read. Please look: https://github.com/openwrt/openwrt/blob/08cc6bc6649f0f184868bd950e38c6107abba0ae/target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-c200-v1.dts#L251

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	reg = <0x00000000 0x20000000>;
+};
+
+chosen {
+	bootargs = "console=ttyS0,115200n8 earlyprintk";
+	stdout-path = &uart0;
+};
+
+mbus {
+	pcie-controller {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:22]: Please use aliases where is possible.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+			gpios = <&gpio1 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+};
+
+&nand {
+	status = "okay";
+
+	partition@0 {
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:157]: Please wrap partitions. Like this: https://github.com/openwrt/openwrt/blob/ed1217eac188b8d4649730b8cf6d35eac6dcd1a8/target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ea3500.dts#L145

File: target/linux/kirkwood/image/Makefile [hunk 6]
```diff
@@ -184,6 +184,25 @@ define Device/iom_ix2-200
 endef
 TARGET_DEVICES += iom_ix2-200
 
+define Device/iom_ix4-200d
+  DEVICE_VENDOR := Iomega
+  DEVICE_MODEL := StorCenter ix4-200d
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-fs-ext4 \
+	kmod-gpio-button-hotplug kmod-mvsdio kmod-hwmon-adt7475 \
+	kmod-spi-gpio kmod-gpio-nxp-74hc164
+  SUPPORTED_DEVICES += marvell,rd88f6281-a
```
Comment (CHKDSK88) [inline target/linux/kirkwood/image/Makefile:193]: why?

File: target/linux/kirkwood/image/Makefile [hunk 7]
```diff
@@ -184,6 +184,25 @@ define Device/iom_ix2-200
 endef
 TARGET_DEVICES += iom_ix2-200
 
+define Device/iom_ix4-200d
+  DEVICE_VENDOR := Iomega
+  DEVICE_MODEL := StorCenter ix4-200d
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-fs-ext4 \
```
Comment (CHKDSK88) [inline target/linux/kirkwood/image/Makefile:190]: Please keep alphabetical order.

File: target/linux/kirkwood/image/Makefile (hunk 6 above)
Comment (CHKDSK88) [reply target/linux/kirkwood/image/Makefile:193]: Why `SUPPORTED_DEVICES ` is added? It is used for legacy purposes.

File: target/linux/kirkwood/base-files/etc/board.d/02_network [hunk 8]
```diff
@@ -37,6 +37,9 @@ kirkwood_setup_interfaces()
 	endian,4i-edge-200)
 		ucidef_set_interface_lan "port1 port2 port3 port4 eth1"
 		;;
+	iom,ix4-200d)
+		ucidef_set_interfaces_lan_wan "eth0" "eth1"
```
Comment (CHKDSK88) [inline target/linux/kirkwood/base-files/etc/board.d/02_network:41]: I'm not sure if NAS should be configured like router. Maybe `ucidef_set_interface_lan "eth0 eth1" "dhcp"` will be better?

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts (hunk 4 above)
Comment (CHKDSK88) [reply target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:22]: Please look here: https://github.com/openwrt/openwrt/blob/a872570c043f5beee21c3af695f1a5a77f7eab04/target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-nsa310s.dts#L260

You can do it simpler:
```
&pciec {
	status = "okay";
};

&pcie0 {
	status = "okay";
};
```

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 9]
```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "kirkwood.dtsi"
+#include "kirkwood-6281.dtsi"
+
+/ {
+	model = "Iomega ix4-200d";
+	compatible = "iom,ix4-200d", "marvell,rd88f6281-a", "marvell,rd88f6281", "marvell,kirkwood-88f6281", "marvell,kirkwood";
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:9]: ` "marvell,rd88f6281-a", "marvell,rd88f6281",` should be dropped.

File: target/linux/kirkwood/image/Makefile (hunk 6 above)
Comment (CHKDSK88) [reply target/linux/kirkwood/image/Makefile:193]: And that is wrong.  Please clean this https://github.com/openwrt/openwrt/pull/12543/files#r1187088402 and `SUPPORTED_DEVICES ` line. First sysupgrade must be with `-F` and next should be work normal.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+ocp@f1000000 {
+	i2c@11000 {
+		status = "okay";
+
+		adt7473@2e {
+			compatible = "adi,adt7473";
+			reg = <0x2e>;
+			};
```
Comment (CHKDSK88) [inline target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:35]: One more nitpick: Ident ;)

Comment (CHKDSK88) [reply target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts]: Yep, only `};`.

File: target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+			gpios = <&gpio_spi 5 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_white: led-power-white {
+			label = "power:white";
+			gpios = <&gpio_spi 7 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+};
```
[the remark this answers is not in the document]
Comment (CHKDSK88) [reply target/linux/kirkwood/files/arch/arm/boot/dts/kirkwood-ix4-200d.dts:129]: You,re right. Indentation is only improper. All subnodes of `/ {`.

**Comment by CHKDSK88** [comment]
> @CHKDSK88 anything I can do to help this get merged? Hope to get this in before 23.x branches smiley

@sanderpqr
Please fix ident in `/ {` node and squash all changes into single commit. After that please push branch with `force` option.

**Pull request closed without merging**